### DISCUSSION TITLE
Feature Request: Vi-alike Interface

### DISCUSSION THREAD

**Discussion by eeriemyxi** [ayn2op/discordo#652] [Ideas]
What the title really means is to basically, instead of relying on another vocabulary of combinations of several hotkeys (which are mostly not at all cozy to press; for example, for the CTRL key, you'd need to put some effort to get back to a state when you could type again), maybe just introduce a vi-alike interface. 

Right now pressing ESC clears the input, however the same thing, on my machine is also achieved by pressing CTRL + U; so that's a duplicated hotkey. What we can do instead is that, when we press ESC, it'd put a hint in the input field that we are currently in "Normal Mode" and then make use of all the keys which are easily accessible to control the UI elements, and when we type `:` in the normal, we will be greeted with a command line (we can just use the input field for this) where we can type ":insert" to go back to "Insert Mode".

The benefits of this implementation:
- Super easy and cozy keyboard control of the UI.
- Very easy to use "plugins" that you have planned (like a plugin to find servers and messages stored in cache; another to view avatars of servers and users; another to download the hinted image and convert it to a ASCII and then display it on the screen, this can be achieved by Go libraries like [this](https://github.com/qeesung/image2ascii#library-usage); another to use `jj` for QWERTY users, or `nn` for Colemak users to activate "Normal Mode" in "Insert Mode").
- Very easy to define custom commands.

**Comment by r3k2** [comment]
Yes please, the reason I use all my software in CLI/TUI is to avoid the mouse, speed and I can use my new gaming pc the same I use my old laptops, my window manager is Wayland/sway. Sucks to have to go grab the mouse for 1 app, and yes, I use Nyxt also a browser with native Emacs/vim ;) I mention this because someone already told me some time a go, "what about for browser, you have to use the mouse" :)  cheers

**Comment by BouleJaune** [comment]
I agree
Any TUI software should at least offer the possibility to use vi-like bindings
Someone wanting a tui version of discord has surely a high probability of using vim/nvim anyway ...

**Comment by ZmoleCristian** [comment]
This would be very much appreciated !

**Comment by nyxmeowmeow** [comment]
did this functionality get removed? i cant find anything like "insert" or "normal" in the current src, and the keybinds from https://github.com/ayn2op/discordo/pull/372 dont do anything.

**Comment by ossenthusiast** [comment]
> did this functionality get removed? i cant find anything like "insert" or "normal" in the current src, and the keybinds from [#372](https://github.com/ayn2op/discordo/pull/372) dont do anything.

Imo, discordo doesnt really benefit from the modal editing experience as much as Vim does.

**Comment by nyxmeowmeow** [comment]
youre probably right in that its not any more efficient for something like this, but i still prefer the modal approach. im more used to it and im not really a fan having to hold down modifiers.
